File: wpilibc/src/main/native/cpp/simulation/REVPHSim.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "frc/simulation/REVPHSim.h"

#include <memory>

#include <hal/simulation/REVPHData.h>

#include "frc/SensorUtil.h"

using namespace frc;
using namespace frc::sim;

REVPHSim::REVPHSim() : PneumaticsBaseSim{SensorUtil::GetDefaultREVPHModule()} {}

REVPHSim::REVPHSim(int module) : PneumaticsBaseSim{module} {}

REVPHSim::REVPHSim(const PneumaticsBase& pneumatics)
    : PneumaticsBaseSim{pneumatics} {}

std::unique_ptr<CallbackStore> REVPHSim::RegisterInitializedCallback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      m_index, -1, callback, &HALSIM_CancelREVPHInitializedCallback);
  store->SetUid(HALSIM_RegisterREVPHInitializedCallback(
      m_index, &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

bool REVPHSim::GetInitialized() const {
  return HALSIM_GetREVPHInitialized(m_index);
}

void REVPHSim::SetInitialized(bool solenoidInitialized) {
  HALSIM_SetREVPHInitialized(m_index, solenoidInitialized);
}

std::unique_ptr<CallbackStore> REVPHSim::RegisterSolenoidOutputCallback(
    int channel, NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      m_index, channel, -1, callback,
      &HALSIM_CancelREVPHSolenoidOutputCallback);
  store->SetUid(HALSIM_RegisterREVPHSolenoidOutputCallback(
      m_index, channel, &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

bool REVPHSim::GetSolenoidOutput(int channel) const {
  return HALSIM_GetREVPHSolenoidOutput(m_index, channel);
}

void REVPHSim::SetSolenoidOutput(int channel, bool solenoidOutput) {
  HALSIM_SetREVPHSolenoidOutput(m_index, channel, solenoidOutput);
}

std::unique_ptr<CallbackStore> REVPHSim::RegisterCompressorOnCallback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      m_index, -1, callback, &HALSIM_CancelREVPHCompressorOnCallback);
  store->SetUid(HALSIM_RegisterREVPHCompressorOnCallback(
      m_index, &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

bool REVPHSim::GetCompressorOn() const {
  return HALSIM_GetREVPHCompressorOn(m_index);
}

void REVPHSim::SetCompressorOn(bool compressorOn) {
  HALSIM_SetREVPHCompressorOn(m_index, compressorOn);
}

std::unique_ptr<CallbackStore> REVPHSim::RegisterCompressorConfigTypeCallback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      m_index, -1, callback, &HALSIM_CancelREVPHCompressorConfigTypeCallback);
  store->SetUid(HALSIM_RegisterREVPHCompressorConfigTypeCallback(
      m_index, &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

int REVPHSim::GetCompressorConfigType() const {
  return HALSIM_GetREVPHCompressorConfigType(m_index);
}

void REVPHSim::SetCompressorConfigType(int compressorConfigType) {
  HALSIM_SetREVPHCompressorConfigType(
      m_index,
      static_cast<HAL_REVPHCompressorConfigType>(compressorConfigType));
}

std::unique_ptr<CallbackStore> REVPHSim::RegisterPressureSwitchCallback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      m_index, -1, callback, &HALSIM_CancelREVPHPressureSwitchCallback);
  store->SetUid(HALSIM_RegisterREVPHPressureSwitchCallback(
      m_index, &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

bool REVPHSim::GetPressureSwitch() const {
  return HALSIM_GetREVPHPressureSwitch(m_index);
}

void REVPHSim::SetPressureSwitch(bool pressureSwitch) {
  HALSIM_SetREVPHPressureSwitch(m_index, pressureSwitch);
}

std::unique_ptr<CallbackStore> REVPHSim::RegisterCompressorCurrentCallback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      m_index, -1, callback, &HALSIM_CancelREVPHCompressorCurrentCallback);
  store->SetUid(HALSIM_RegisterREVPHCompressorCurrentCallback(
      m_index, &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

double REVPHSim::GetCompressorCurrent() const {
  return HALSIM_GetREVPHCompressorCurrent(m_index);
}

void REVPHSim::SetCompressorCurrent(double compressorCurrent) {
  HALSIM_SetREVPHCompressorCurrent(m_index, compressorCurrent);
}

uint8_t REVPHSim::GetAllSolenoidOutputs() const {
  uint8_t ret = 0;
  HALSIM_GetREVPHAllSolenoids(m_index, &ret);
  return ret;
}

void REVPHSim::SetAllSolenoidOutputs(uint8_t outputs) {
  HALSIM_SetREVPHAllSolenoids(m_index, outputs);
}

void REVPHSim::ResetData() {
  HALSIM_ResetREVPHData(m_index);
}

File: wpilibj/src/main/java/org/wpilib/simulation/REVPHSim.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.wpilibj.simulation;

import edu.wpi.first.hal.simulation.NotifyCallback;
import edu.wpi.first.hal.simulation.REVPHDataJNI;
import edu.wpi.first.wpilibj.PneumaticHub;
import edu.wpi.first.wpilibj.SensorUtil;

/** Class to control a simulated PneumaticHub (PH). */
public class REVPHSim extends PneumaticsBaseSim {
  /** Constructs for the default PH. */
  public REVPHSim() {
    super(SensorUtil.getDefaultREVPHModule());
  }

  /**
   * Constructs from a PH module number (CAN ID).
   *
   * @param module module number
   */
  public REVPHSim(int module) {
    super(module);
  }

  /**
   * Constructs from a PneumaticHum object.
   *
   * @param module PCM module to simulate
   */
  public REVPHSim(PneumaticHub module) {
    super(module);
  }

  /**
   * Check whether the closed loop compressor control is active.
   *
   * @return config type
   */
  public int getCompressorConfigType() {
    return REVPHDataJNI.getCompressorConfigType(m_index);
  }

  /**
   * Turn on/off the closed loop control of the compressor.
   *
   * @param compressorConfigType compressor config type
   */
  public void setCompressorConfigType(int compressorConfigType) {
    REVPHDataJNI.setCompressorConfigType(m_index, compressorConfigType);
  }

  /**
   * Register a callback to be run whenever the closed loop state changes.
   *
   * @param callback the callback
   * @param initialNotify whether the callback should be called with the initial value
   * @return the {@link CallbackStore} object associated with this callback.
   */
  public CallbackStore registerCompressorConfigTypeCallback(
      NotifyCallback callback, boolean initialNotify) {
    int uid = REVPHDataJNI.registerCompressorConfigTypeCallback(m_index, callback, initialNotify);
    return new CallbackStore(m_index, uid, REVPHDataJNI::cancelCompressorConfigTypeCallback);
  }

  @Override
  public boolean getInitialized() {
    return REVPHDataJNI.getInitialized(m_index);
  }

  @Override
  public void setInitialized(boolean initialized) {
    REVPHDataJNI.setInitialized(m_index, initialized);
  }

  @Override
  public CallbackStore registerInitializedCallback(NotifyCallback callback, boolean initialNotify) {
    int uid = REVPHDataJNI.registerInitializedCallback(m_index, callback, initialNotify);
    return new CallbackStore(m_index, uid, REVPHDataJNI::cancelInitializedCallback);
  }

  @Override
  public boolean getCompressorOn() {
    return REVPHDataJNI.getCompressorOn(m_index);
  }

  @Override
  public void setCompressorOn(boolean compressorOn) {
    REVPHDataJNI.setCompressorOn(m_index, compressorOn);
  }

  @Override
  public CallbackStore registerCompressorOnCallback(
      NotifyCallback callback, boolean initialNotify) {
    int uid = REVPHDataJNI.registerCompressorOnCallback(m_index, callback, initialNotify);
    return new CallbackStore(m_index, uid, REVPHDataJNI::cancelCompressorOnCallback);
  }

  @Override
  public boolean getSolenoidOutput(int channel) {
    return REVPHDataJNI.getSolenoidOutput(m_index, channel);
  }

  @Override
  public void setSolenoidOutput(int channel, boolean solenoidOutput) {
    REVPHDataJNI.setSolenoidOutput(m_index, channel, solenoidOutput);
  }

  @Override
  public CallbackStore registerSolenoidOutputCallback(
      int channel, NotifyCallback callback, boolean initialNotify) {
    int uid =
        REVPHDataJNI.registerSolenoidOutputCallback(m_index, channel, callback, initialNotify);
    return new CallbackStore(m_index, channel, uid, REVPHDataJNI::cancelSolenoidOutputCallback);
  }

  @Override
  public boolean getPressureSwitch() {
    return REVPHDataJNI.getPressureSwitch(m_index);
  }

  @Override
  public void setPressureSwitch(boolean pressureSwitch) {
    REVPHDataJNI.setPressureSwitch(m_index, pressureSwitch);
  }

  @Override
  public CallbackStore registerPressureSwitchCallback(
      NotifyCallback callback, boolean initialNotify) {
    int uid = REVPHDataJNI.registerPressureSwitchCallback(m_index, callback, initialNotify);
    return new CallbackStore(m_index, uid, REVPHDataJNI::cancelPressureSwitchCallback);
  }

  @Override
  public double getCompressorCurrent() {
    return REVPHDataJNI.getCompressorCurrent(m_index);
  }

  @Override
  public void setCompressorCurrent(double compressorCurrent) {
    REVPHDataJNI.setCompressorCurrent(m_index, compressorCurrent);
  }

  @Override
  public CallbackStore registerCompressorCurrentCallback(
      NotifyCallback callback, boolean initialNotify) {
    int uid = REVPHDataJNI.registerCompressorCurrentCallback(m_index, callback, initialNotify);
    return new CallbackStore(m_index, uid, REVPHDataJNI::cancelCompressorCurrentCallback);
  }

  @Override
  public void resetData() {
    REVPHDataJNI.resetData(m_index);
  }
}
